testsuite/tests/lexical_envs/named_envs/expected_concrete_syntax.lkt

```diff
@@ -0,0 +1,278 @@
+}
+
+class DeclarativePart : FooNode {
+    @parse_field decls : ASTList[FooNode]
+}
+
+class PrivatePart : DeclarativePart {
+}
+
+class PublicPart : DeclarativePart {
+}
+
+## Top-level list of declaration in a subprogam body.
+class SubpBodyDecls : DeclarativePart {
+}
+
+class DeclareBlock : FooNode {
+    @parse_field decls : DeclarativePart
+    @parse_field stmts : ASTList[FooNode]
+}
+
+class IncompleteTypeDecl : FooNode {
+    @parse_field name : Identifier
+}
+
+@abstract class Name : FooNode {
+
+    ## Innermost suffix for this name.
+    fun base_name (): Symbol = match node {
+        case id : Identifier => id.symbol
+        case dn : DottedName => dn.suffix.base_name()
+    }
+
+    ## Absolute name of the scope that this name defines, assuming that
+    ## ``prefix`` is the implicit prefix for that name.
+    fun scope_name (): Array[Symbol] = match node {
+        case id : Identifier => [id.symbol]
+        case dn : DottedName => dn.prefix.scope_name() & ([dn.suffix.symbol])
+    }
+
+    ## Absolute name of the scope that defines this name, assuming that
+    ## ``prefix`` is the implicit prefix for that name.
+    fun parent_scope_name (): Array[Symbol] = match node {
+        case _ : Identifier => null
+        case dn : DottedName => dn.prefix.scope_name()
+    }
+
+    ## Innermost scope for this name. The lookup starts from the given
+    ## environment.
+    fun scope (from_env : LexicalEnv): LexicalEnv =
+    node.resolve(from_env).children_env()
+
+    ## Return the scope that contains the declaration that this name defines.
+    fun parent_scope (from_env : LexicalEnv): LexicalEnv = match node {
+        case _ : Identifier => from_env
+        case dn : DottedName => dn.prefix.scope(from_env)
+    }
+
+    ## Return the list of declarations that define this name. The lookup
+    ## starts from the given environment.
+    fun resolve (from_env : LexicalEnv): FooNode =
+    node.parent_scope(from_env).get_first(node.base_name())
+}
+
+class DottedName : Name {
+    @parse_field prefix : Name
+    @parse_field suffix : Identifier
+}
+
+class Identifier : Name implements TokenNode {
+}
+
+class PackageBody : FooNode {
+    @parse_field name : Name
+    @parse_field decls : DeclarativePart
+
+    fun self_name (): Name = node.name
+
+    ## Assuming this PackageBody cannot define a named environment, return the
+    ## environment in which this declaration should be registered.
+    fun get_initial_env (): LexicalEnv = {
+        val pkg_decl = node.parent.children_env().get_first(node.name.base_name()).as[PackageDecl];
+
+        if (pkg_decl.private_part.is_null) then pkg_decl.children_env() else pkg_decl.private_part.children_env()
+    }
+
+    ## Return the PackageDecl node corresponding to this PackageBody.
+    @export fun decl_part (): PackageDecl = {
+        val env = node.parent.children_env();
+
+        node.name.resolve(env).as[PackageDecl]
+    }
+}
+
+class PackageDecl : FooNode {
```

Comment:
I’m not sure what you mean: all tests dumping Lkt sources unparse at least the grammar and the nodes. Here, we just happen to unparse the lexer in addition, so that we can have in this testcase a syntax that really looks like Ada.

Reply:
OK